testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,297 @@
+		goto error;
+	}
+
+	written = write(fd, tc.cpattern, tc.cpsize);
+	if (written < (size_t)tc.cpsize) {
+		tst_brk(TINFO | TERRNO, "write(fd, tc.pattern, ...)");
+		goto error;
+	}
+
+	rval = close(fd);
+	if (rval == -1) {
+		tst_brk(TINFO | TERRNO, "close(fd)");
+		goto error;
+	}
+
+	tc.cpsize = 0;
+	return;
+error:
+	tst_brk(TBROK,
+		"COULD NOT RESTORE "CORE_PATTERN"! It used to be '%s'",
+		tc.cpattern);
+}
+
+static void cleanup(void)
+{
+	restore_cpattern();
+
+	if (tc.fmem != NULL && munmap(tc.fmem, tc.fmemsize))
+		tst_brk(TBROK | TERRNO, "munmap(tc.fmem)");
+	tc.fmem = NULL;
+
+	if (tc.fd > 0 && close(tc.fd))
+		tst_brk(TBROK | TERRNO, "close(tc.fd)");
+
+	if (tc.dfd > 0 && close(tc.dfd))
+		tst_brk(TBROK | TERRNO, "close(tc.dfd)");
+}
+
+static int find_sequence(int pid)
+{
+	char c, expectc = 'x';
+	int ycount = 0;
+	char dumpname[256];
+
+	snprintf(dumpname, 256, "dump-%d", pid);
+	tst_res(TINFO, "Dump file should be %s", dumpname);
+	if (access(dumpname, F_OK))
+		tst_brk(TCONF | TERRNO,
+			"Dump file was not found; assuming this is because "
+			"core dumps are disabled and not a bug");
+	tc.dfd = SAFE_OPEN(dumpname, O_RDONLY);
+
+	while (SAFE_READ(0, tc.dfd, &c, 1)) {
+		switch (c) {
+		case 'x':
+			ycount = 0;
+			expectc = 'y';
+			break;
+		case 'y':
+			if (expectc == 'y') {
+				ycount++;
+			} else {
+				expectc = 'x';
+				break;
+			}
+
+			if (ycount == tc.ycount)
+				expectc = 'z';
+			break;
+		case 'z':
+			if (expectc == 'z')
+				return 1;
+		default:
+			expectc = 'x';
+		}
+	}
+	return 0;
+}
+
+static void run_child(int advice)
+{
+	char *advstr =
+		advice == MADV_DONTDUMP ? "MADV_DONTDUMP" : "MADV_DODUMP";
+
+	if (madvise(tc.fmem, tc.fmemsize, advice) == -1) {
+		tst_res(TFAIL | TERRNO,
+			"madvise(%p, %lu, %s) = -1",
+			tc.fmem,
+			tc.fmemsize,
+			advstr);
+		return;
+	}
+	abort();
+}
+
+static void run(void)
+{
+	int i, status;
+	pid_t pid;
+	char *fmemc;
+
+	tc.fmem = SAFE_MMAP(NULL,
+			    tc.fmemsize,
+			    PROT_READ | PROT_WRITE,
+			    MAP_ANONYMOUS | MAP_PRIVATE,
+			    -1,
+			    0);
+	/*Write a generated character sequence to the mapped memory,
+	  which we later look for in the core dump.*/
+	fmemc = (char *)tc.fmem;
+	*fmemc = 'x';
+	for (i = 0; i < tc.ycount; i++)
+		fmemc[i + 1] = 'y';
+	fmemc[++i] = 'z';
+
+	pid = SAFE_FORK();
+	if (pid == 0) {
+		run_child(MADV_DONTDUMP);
+	}
+
+	SAFE_WAITPID(pid, &status, 0);
+	if (!(WIFSIGNALED(status) && WCOREDUMP(status)))
+		tst_brk(TCONF, "No core dump produced after abort()");
+
+	if (find_sequence(pid)) {
+		tst_res(TFAIL, "Found sequence in dump despite MADV_DONTDUMP");
+		tst_res(TFAIL, "MADV_DODUMP is an automatic fail");
+		return;
+	} else {
+		tst_res(TPASS, "madvise(..., MADV_DONTDUMP");
+	}
+
+	pid = SAFE_FORK();
+	if (pid == 0) {
+		run_child(MADV_DODUMP);
+	}
+
+	SAFE_WAITPID(pid, &status, 0);
+	if (!(WIFSIGNALED(status) && WCOREDUMP(status)))
+		tst_brk(TCONF, "No core dump produced after abort()");
+
+	restore_cpattern();
+
```

Comment:
Again, the ```run()``` function can be called in a loop, so you should either set and restore the pattern at the start/end of the ```run()``` function or in ```setup()```/```cleanup()``` only.